messaging/src/main/java/org/axonframework/common/infra/FilesystemStyleComponentDescriptor.java

```diff
@@ -0,0 +1,377 @@
+ * <p>
+ * Each component is assigned a path in the virtual filesystem. References to components that have already been
+ * described are represented as symbolic links, indicated with the {@code -> /path/to/component} syntax, preventing
+ * infinite recursion in circular references.
+ * <p>
+ * Example output:
+ * <pre>
+ * /
+ * ├── systemComponent/
+ * │   ├── _ref: 123512
+ * │   ├── _type: EventProcessor
+ * │   ├── name: my-processor
+ * │   └── configuration/
+ * │       ├── _ref: 123513
+ * │       ├── _type: ProcessorConfiguration
+ * │       ├── batchSize: 100
+ * │       └── processor -> /systemComponent
+ * </pre>
+ *
+ * @author Mateusz Nowak
+ * @since 5.0.0
+ */
+public class FilesystemStyleComponentDescriptor implements ComponentDescriptor {
+
+    private static final String ROOT_PATH = "/";
+    private static final String PATH_SEPARATOR = "/";
+
+    private final Map<Object, String> componentPaths;
+    private final Map<String, Object> properties;
+    private final String currentPath;
+
+    /**
+     * Constructs a new {@code FilesystemComponentDescriptor} as the root of the hierarchy.
+     */
+    public FilesystemStyleComponentDescriptor() {
+        this(new IdentityHashMap<>(), ROOT_PATH);
+    }
+
+    /**
+     * Private constructor used for creating nested descriptors that share the component paths map.
+     *
+     * @param componentPaths Map containing paths to already processed components.
+     * @param currentPath    The current path in the virtual filesystem.
+     */
+    private FilesystemStyleComponentDescriptor(Map<Object, String> componentPaths, String currentPath) {
+        this.componentPaths = componentPaths;
+        this.currentPath = currentPath;
+        this.properties = new LinkedHashMap<>();
+    }
+
+    @Override
+    public void describeProperty(@Nonnull String name, Object object) {
+        if (object instanceof DescribableComponent component) {
+            describeComponent(name, component);
+        } else {
+            properties.put(name, object);
+        }
+    }
+
+    private void describeComponent(String name, DescribableComponent component) {
+        var componentPath = currentPathChild(name);
+
+        boolean componentSeenAlready = componentPaths.containsKey(component);
+        if (componentSeenAlready) {
+            var existingPath = componentPaths.get(component);
+            properties.put(name, new SymbolicLink(existingPath));
+            return;
+        }
+
+        componentPaths.put(component, componentPath);
+        properties.put(name, componentDescriptor(component, componentPath));
+    }
+
+    private String currentPathChild(String name) {
+        return currentPath.equals(ROOT_PATH)
+                ? currentPath + name
+                : currentPath + PATH_SEPARATOR + name;
+    }
+
+    @Override
+    public void describeProperty(@Nonnull String name, Collection<?> collection) {
+        if (collection == null) {
```

Comment:
Nit: shouldn't we do a null check on the `describeProperty(String, Object)` method too?

Reply:
Not needed, it tested. For collection/map is needed, because we iterate over them (cause NullPointerException). For object, it works OK.

Reply:
But of course! Thanks for clarifying.